### ISSUE TITLE
Tracking sheet issues

### ISSUE THREAD

**Issue by LynnMcRae** [sul-dlss/argo#690]
From Astrid:

> I tried again to resister some of the object I'd been working with earlier, and when I tried to print the 
> tracking sheets I got an error message (I copied it, but it got erased when I copied the feedback button 
> email address.)  Then I tried again, and the tracking sheets are all blank.  

and 

> After clicking the tracking sheet button, the registration page was no longer showed the lines that were entered for each item -- some had worked and some hadn't, so 
> in order to tell which items successfully registered, I started searching each of them by ckey (7 items) and found a strange status for some (no workflow?) They are not showing as registered, but the druid was created. 
> https://argo.stanford.edu/catalog/druid:jm811sw5601

![screen shot 2016-03-18 at 2 58 28 pm](https://cloud.githubusercontent.com/assets/1824732/13894336/d98cd058-ed23-11e5-8d0f-d397cd8163de.png)

John did some experimentation and says he could get one druid to work but not > 1

**Comment by jkeck** [comment]
@LynnMcRae am I to understand that this ticket is adding more additional non-media related work to our queue?

**Comment by LynnMcRae** [comment]
Unfortunately, anything that actually broke with the release is the sort of thing we'd always try to address in and among other work as a production system problem.  But we are nonetheless trying to wrap up this week's round, get it out, then evaluate the urgency of what is left. The labs rely on the tracking sheets for their workflow depending on the type of materials being digitized, so it is likely serious.

**Comment by tingulfsen** [comment]
@LynnMcRae can you explain how to reproduce this bug? I have not worked with tracking sheets before and it's unclear to me what is required to get to them.

**Comment by LynnMcRae** [comment]
I think it's as simple as doing a successful registration run, so you got a bunch of new druids on the registration screen, then you click on the Tracking Sheets link and out pops (downloads) a file for you to print.  But having never done one, I defer to you @blalbrit for more accurate information.

**Comment by blalbrit** [comment]
@LynnMcRae - yep that's right. Register objects in registration form. Once registration is complete, click the "tracking sheets" button which downloads pdf.

**Comment by LynnMcRae** [comment]
someday in the future we'll support tracking sheets as an out put from a search result or as a new Bulk Action for a list of druids.

**Comment by blalbrit** [comment]
it's currently working in stage

**Comment by blalbrit** [comment]
It's also currently working in prod. @LynnMcRae is it possible this was a false error reported by either slow indexing or the Fedora failures we saw on Friday?

**Comment by LynnMcRae** [comment]
roger that @blalbrit have the labs tried it since or are they waiting to hear from us?

**Comment by jmartin-sul** [comment]
this was astrid's feedback message

> Got back to the error message:
> 
> File not found
> 
> Firefox can't find the file at https://argo.stanford.edu/registration/tracksheet?druid[]=&druid[]=bk326tn9935&druid[]=gc042fb1556&druid[]=&druid[]=&druid[]=ft623fr7194&druid[]=xv847sp9691&druid[]=nv163vg0156&name=Memorial+Library+of+Music&sequence=1.
>
>    Check the file name for capitalization or other typing errors.
>    Check to see if the file was moved, renamed or deleted.

i believe you can hit the tracking sheet URL whenever you want for any object, there's just no link or redirect to it in the application other than via item registration.  note that this single druid link should work even though the above doesn't (and IIRC, it used to support specifying multiple druids, but that should be testable on argo-test.stanford.edu):  https://argo.stanford.edu/registration/tracksheet?druid[]=nv163vg0156&name=Memorial+Library+of+Music&sequence=1

**Comment by blalbrit** [comment]
The first URL in @jmartin-sul 's example doesn't work because of three blank druid parameters (the first druid[]= after "tracksheet?" and two others further along. Once those three pieces of bad data were removed, the file downloaded as expected. See https://argo.stanford.edu/registration/tracksheet?druid[]=bk326tn9935&druid[]=gc042fb1556&druid[]=ft623fr7194&druid[]=xv847sp9691&druid[]=nv163vg0156&name=Memorial+Library+of+Music&sequence=1

**Comment by blalbrit** [comment]
Looking at https://argo.stanford.edu/report?f%5Bnonhydrus_collection_title_ssim%5D%5B%5D=Memorial+Library+of+Music , it looks like there are 3 objects in processing status = "unknown status" that likely correspond to the three missing druid parameters above. My guess is that they failed to get the registration workflow applied at time of registration (b/c of the other issues we were seeing on Friday?) - leading to the broken tracking sheet link seen above.

This might be an edge case, but maybe we can talk about what preferred failure behavior would be for:

- objects that are created in DOR but don't get status "registered"

- tracking sheet lists items where the system can't find one or more objects


@LynnMcRae and I are following up with Astrid to see if the issue has cleared up since the fixes applied on Friday.

**Comment by jmartin-sul** [comment]
ah, yeah, i'd noticed the first blank param and tried removing it, but not the subsequent ones.

for the latter point (tracking sheet item lists where the system can't find one or more objects, or where the list has blank druids):  i'd propose that we give tracking sheets for what we can, and an error message about blanks or failures, though i'm not sure where that error message should be presented.  seems like it'd be nice to present that info on whatever page redirected to the tracking sheet link (e.g., maybe instead of redirecting automatically in this case, a warning could be presented with a link to proceed).

**Comment by LynnMcRae** [comment]
@jmartin-sul is tracking sheets applying some status criteria as to what it will print? Do you know the reason for that? I think we should print a sheet regardless of the state of the object once we have a druid unless there's a reason in the tracking sheet itself (but the user could at least see and discard such sheets if needed).  Looking for information on this only and will do a separate ticket for a change to the tracking sheet logic. The immediate issue seems to be resolved sufficiently as long as Fedora stays up :)

**Comment by jmartin-sul** [comment]
@LynnMcRae: i don't think it does.  what it does do is look up the object in solr, and if it's not in solr, it tries to retrieve it form fedora and then index it to solr:
1. https://github.com/sul-dlss/argo/blob/v3.0.2/app/controllers/registration_controller.rb#L6-L13
2. https://github.com/sul-dlss/argo/blob/v3.0.2/app/models/track_sheet.rb#L15
3. https://github.com/sul-dlss/argo/blob/v3.0.2/app/models/track_sheet.rb#L108-L121

1 is the controller method that fulfills the request, 2 loops through the list of druids it's given and does the thing that makes sure their indexed, 3 does the indexing check for a given druid.  my guess, without trying this myself on the console or in pry, is that step 3 is failing on line 115 and the 404 is bubbling up.  note that this URL gets the same error that astrid ran into:
https://argo.stanford.edu/registration/tracksheet?druid[]=bk326tn9935&druid[]=gc042fb1556&druid[]=ft623fr7194&druid[]=xv847sp9691&druid[]=blargh&name=Memorial+Library+of+Music&sequence=1

no blanks in there, but `blargh` is not a druid that exists in fedora, so i think that's a pretty good indication that this is a lookup error.  the thing that gives me doubt about that theory is that we don't present the usual 404 in this case.

**Comment by jmartin-sul** [comment]
which is to say, as far as i can tell, the tracking sheet code should generate a sheet for anything that exists in fedora.  it also appears that missing fields should be dealt with gracefully, as it looks like the fields are all checked for existence before being added to the PDF:
https://github.com/sul-dlss/argo/blob/v3.0.2/app/models/track_sheet.rb#L91-L106

**Comment by LynnMcRae** [comment]
ok, so not a status thing at all, thanks.  Should I still do a separate ticket for this, or are you trying to fix the one-bad-druid-spoils-the-bunch effect described here?

**Comment by LynnMcRae** [comment]
Talked to John ... this ticket suffices for future improvements, but taking it out of post-release must-haves

**Comment by andrewjbtw** [comment]
No longer relevant. If there's an initial failure in tracking sheet generation post-registration, the tracking sheet bulk action can be used to generate new sheets for any registered druid.
